onnxruntime/core/platform/windows/env.cc

```diff
@@ -124,52 +147,171 @@ class WindowsEnv : public Env {
+        }
+        kAffinities.push_back(group_id);
+        kAffinities.push_back(processor_mask);
       }
-      // This is the number of logical processors in the current group
-      return sysInfo.dwNumberOfProcessors;
-    }
-    int processorCoreCount = 0;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
-      if (buffer[i].Relationship == RelationProcessorCore) {
-        ++processorCoreCount;
+      group_id++;
+    }
+    if (kAffinities.empty()) {
+      return std::thread::hardware_concurrency() >> 1;
+    }
+    size_t num_of_size_t = std::max(kAffinities.size() * sizeof(KAFFINITY) / sizeof(size_t), 1ULL);
+    affinities.resize(num_of_size_t, 0);
+    memcpy(affinities.data(), kAffinities.data(), kAffinities.size() * sizeof(KAFFINITY));
+    return kAffinities.size() >> 1;
+  }
+
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(int procssor_from, int processor_to) {
+    if (procssor_from > processor_to) {
+      LOGS_DEFAULT(ERROR) << "Processor <from> must be smaller or equal to <to>";
+      return {-1, 0};
+    }
+    int processor_id = 1;
+    int processor_count = 0;
+    KAFFINITY group_id = 0;
+    KAFFINITY processor_mask = 0;
+    for (const auto& group_info : this->group_info_vec_) {
+      for (int i = 0; i < group_info.ActiveProcessorCount; ++i, ++processor_id) {
+        if (processor_id >= procssor_from && processor_id <= processor_to) {
+          processor_mask |= BitOne << i;
+          processor_count += 1;
+        }
+      }
+      // when mask is set, do not visit next group
+      if (processor_mask) {
+        break;
       }
+      group_id++;
     }
-    if (!processorCoreCount)
-      ORT_THROW("Fatal error: 0 count processors from GetLogicalProcessorInformation");
-    return processorCoreCount;
-  }
-
-  std::vector<size_t> GetThreadAffinityMasks() const override {
-    auto generate_vector_of_n = [](int n) {
-      std::vector<size_t> ret(n);
-      std::iota(ret.begin(), ret.end(), 0);
-      return ret;
-    };
-    // Indeed 64 should be enough. However, it's harmless to have a little more.
-    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-    DWORD returnLength = sizeof(buffer);
-    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
-      return generate_vector_of_n(std::thread::hardware_concurrency());
-    }
-    std::vector<size_t> ret;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
-      if (buffer[i].Relationship == RelationProcessorCore) {
-        ret.push_back(buffer[i].ProcessorMask);
+    if (processor_count < processor_to - procssor_from + 1) {
+      LOGS_DEFAULT(ERROR) << "Processor <from> or <to> cross group boundary";
+      return {-1,0};
+    }
+    return {group_id, processor_mask};
+  }
+
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(const std::vector<std::string>& processor_id_strs) {
```

Comment:
Should avoid using std::string directly. We should use std::basic_string<ORTCHAR_T> instead. On Windows it maps to std::wstring and on the other platforms it maps to std::string.  Or, you explicitly say this string is a utf-8 string.

Reply:
Talked to Randy offline. He told me this processor_id_strs string is ASCII only. So either is ok.  Ideally in onnxruntime_c_api.h we should have a consistent policy on this. For example, use wchar_t* for all strings on Windows and char* on Linux.  Now it is case by case. It is easy to make mistakes if someone didn't carefully read the comments.